File: projects/21-bridge-structural-health/backend/processor/transform.py
"""Pure transform: one fog window-aggregate message -> one DynamoDB item."""

import json


def process(message_body):
    data = json.loads(message_body) if isinstance(message_body, str) else message_body
    site_id = data.get("site_id", "span-a")
    return {
        "sensor_type": data["sensor_type"],
        # window_end alone would collide when span-a and span-b flush in
        # the same window, since sensor_type is the DynamoDB partition key
        # -- concatenating site_id disambiguates the two spans.
        "sort_key": f"{data['window_end']}#{site_id}",
        "window_end": data["window_end"],
        "window_start": data["window_start"],
        "site_id": site_id,
        "unit": data.get("unit", ""),
        "count": data["count"],
        "min": data["min"],
        "max": data["max"],
        "avg": data["avg"],
        "latest": data["latest"],
        "alerts": data.get("alerts", []),
    }

File: projects/18-elevator-escalator-fleet-monitoring/backend/processor/transform.js
"use strict";

// sort_key = window_end#site_id keeps tower-a and tower-b from colliding on the same partition key.
function buildSortKey(windowEnd, siteId) {
  return `${windowEnd}#${siteId}`;
}

function toItem(messageBody) {
  const payload = typeof messageBody === "string" ? JSON.parse(messageBody) : messageBody;
  const siteId = payload.site_id || "tower-a";
  return {
    sensor_type: payload.sensor_type,
    sort_key: buildSortKey(payload.window_end, siteId),
    window_end: payload.window_end,
    window_start: payload.window_start,
    site_id: siteId,
    unit: payload.unit || "",
    count: payload.count,
    min: payload.min,
    max: payload.max,
    avg: payload.avg,
    latest: payload.latest,
    alerts: payload.alerts || [],
  };
}

module.exports = { toItem, buildSortKey };
